### PR TITLE
Add RECORD as an allowed type for Relationships

### PR DESCRIPTION
**Pull request by jphui**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
  - [META-19786](https://jira01.corp.linkedin.com:8443/browse/META-19786)
- [x] Tests for the changes have been added/updated (if applicable)
  - n/a
- [x] Docs related to the changes have been added/updated (if applicable)

## Reference code

I used [this blurb](https://jarvis.corp.linkedin.com/codesearch/result/?name=SettingsUtil.java&path=identity%2Fmember-api%2Fsrc%2Fmain%2Fjava%2Fcom%2Flinkedin%2Fidentity%2Fmember&reponame=linkedin-multiproduct%2Fidentity#155) from `identity` MP's `SettingsUtil.java` found through CodeSearch to base my logic:

```
      } else if (field.getType().getType() == DataSchema.Type.RECORD
          && Locale.class.getCanonicalName().equals(field.getType().getUnionMemberKey())) {
        // Serialize the locale object
        return localeToSerializedString((Locale) object);
      } else {
```

### CODE REVIEW

File: docs/what/relationship.md [hunk 1]
```diff
@@ -61,6 +61,7 @@ Below is an example of how a relationship is modeled in PDL. Note that:
 2. Each model is expected to have a `@pairings` annotation that is an array of all allowed source-destination URN pairs.
 3. Unlike entity attributes, there’s no requirement on making all relationship attributes optional since relationships
    do not support partial updates.
+4. Only primitive and RECORD field types are allowed.
```
Comment (yangyangv2) [inline docs/what/relationship.md:64]: I feel a bit risky to open for all RECORDs. Can it be restricted to AuditStamp and at "auditStamp" field?

The consideration is. to support indexing AuditStamp properly, we need to properly set the ES data type for AuditStamp, e.g. 'keywords' for actor, 'Number' for time ..etc. Any random structured data object will be ended with unpredictable indexing options with ES's dynamic mapping.

Comment (yangyangv2) [reply docs/what/relationship.md:64]: Alternatively, we can also honor the current primitive restrictions by flatten the AuditStamp into actor, time fields.  (e.g. aduitActor, auditTime)

File: docs/what/relationship.md [hunk 2]
```diff
@@ -61,6 +61,7 @@ Below is an example of how a relationship is modeled in PDL. Note that:
 2. Each model is expected to have a `@pairings` annotation that is an array of all allowed source-destination URN pairs.
 3. Unlike entity attributes, there’s no requirement on making all relationship attributes optional since relationships
    do not support partial updates.
+4. Only primitive field types are allowed.
```
Comment (yangyangv2) [inline docs/what/relationship.md:64]: thanks for keeping the readme up-to-date!

File: validators/src/main/java/com/linkedin/metadata/validator/ValidationUtils.java [hunk 3]
```diff
@@ -161,6 +162,16 @@ public static List<RecordDataSchema.Field> fieldsUsingInvalidType(@Nonnull Recor
         .collect(Collectors.toList());
   }
 
+  /**
+   * Return true if the field is an AuditStamp of type RECORD.
+   *   META-19786 Add AuditStamp to allowlist to enable addition to BaseRelationship
+   */
+  @Nonnull
+  public static boolean isValidAuditStamp(@Nonnull RecordDataSchema.Field field) {
```
Comment (yangyangv2) [inline validators/src/main/java/com/linkedin/metadata/validator/ValidationUtils.java:170]: can you also add `field.getName().equals("auditStamp")`

("auditStamp" is the attribute name you're going to add into the BaseRelationship)

File: validators/src/main/java/com/linkedin/metadata/validator/ValidationUtils.java [hunk 4]
```diff
@@ -161,6 +162,16 @@ public static List<RecordDataSchema.Field> fieldsUsingInvalidType(@Nonnull Recor
...
+    return field.getType().getUnionMemberKey().equals(AuditStamp.class.getCanonicalName())
+        && getFieldOrArrayItemType(field) == DataSchema.Type.RECORD;
```
Comment (yangyangv2) [inline validators/src/main/java/com/linkedin/metadata/validator/ValidationUtils.java:172]: if type is `AuditStamp`, do we still need to verify if it's `DataSchema.Type.RECORD`?

**Comment by jphui** [comment]
Closing, have decided to split up AuditStamp into its compositional parts (which are primitives) instead in order to honor historical reasons why it was prevented in the first place and to hedge against unforeseen tech debt / issues.

**Pull request closed without merging**